dcos_launch/terraform.py

```diff
@@ -0,0 +1,144 @@
+        YAML Configuration" section here: https://github.com/mesosphere/terraform-dcos-enterprise/tree/master/aws
+        """
+        with open(self.cluster_profile_path, 'w') as file:
+            for k, v in self.config['terraform_config'].items():
+                file.write(k + ' = ')
+                if type(k) is dict:
+                    file.write('<<EOF\n{}\nEOF\n'.format(yaml.dump(v)))
+                else:
+                    file.write('"{}"\n'.format(v))
+
+        # TODO ssh key helpers
+        subprocess.run(['terraform', 'init', '-from-module', module], cwd=self.init_dir, check=True,
+                       stderr=subprocess.STDOUT)
+        subprocess.run(['terraform', 'apply', '-auto-approve', '-var-file', self.cluster_profile_path],
+                       cwd=self.init_dir, check=True, stderr=subprocess.STDOUT)
+        return self.config
+
+    def wait(self):
+        """ Nothing to do here because unlike the other launchers, create() runs and also waits for a subprocess to
+        finish instead of just sending an http request to a provider.
+        """
+        pass
+
+    def delete(self):
+        subprocess.run(['terraform', 'destroy', '-force', '-var-file', self.cluster_profile_path], cwd=self.init_dir,
+                       check=True, stderr=subprocess.STDOUT)
+        shutil.rmtree(self.init_dir, ignore_errors=True)
+
+    def describe(self) -> dict:
+        """ Output example
+        Outputs:
+
+        Bootstrap Public IP Address = 35.230.27.229
+        Master ELB Address = 35.230.34.40
+        Mesos Agent Public IP = [
+            35.230.112.212,
+            35.230.68.63
+        ]
+        Mesos Master Public IP = [
+            35.230.48.204
+        ]
+        Mesos Public Agent Public IP = [
+            35.197.122.209
+        ]
+        Public Agent ELB Address = 35.230.52.188"""
+        result = subprocess.run(['terraform', 'show'], cwd=self.init_dir, check=True, stdout=subprocess.PIPE)
+        # crop the output to speed up regex
+        info = result.stdout[result.stdout.find('Outputs:\n\n'):]
+
+        ip_regex = '\d.+'
+        ip_list_regex = '\[[^\[]+\]'
+
+        def get_ips(*arg) -> list:
+            ips = []
+            for prefix in arg:
+                prefix += ' = '
+                m = re.search('{}{}'.format(prefix, ip_regex), info)
+                if m:
+                    ips.append(m.group(0)[len(prefix):])
+                else:
+                    m = re.search('{}{}'.format(prefix, ip_list_regex), info)
+                    if m:
+                        # remove prefix
+                        s = m.group(0)[len(prefix):]
+                        # remove whitespace
+                        s = "".join(s.split())
+                        # remove brackets
+                        s = s[1:-1]
+                        ips += s.split(',')
+            return ips
+
+        def convert_to_describe_format(ips: list) -> list:
```

Comment:
same as above; I would prefer a top-level private method to a double-indented inner function

Reply:
Done